python/tank/platform/software_launcher.py

```diff
@@ -205,72 +221,185 @@ def display_name(self):
-                              to search for. If set to None or [], search
-                              for all versions. A version string is
-                              DCC-specific but could be something
-                              like "2017", "6.3v7" or "1.2.3.52"
-        :returns: List of :class:`SoftwareVersion` instances
+    @property
+    def versions(self):
         """
-        raise NotImplementedError
+        A list of versions limiting executable discovery.
+
+        Example::
+
+            ["15.5.324", "16.0.1.322"]
+        """
+        return self._versions
+
+    ##########################################################################################
+    # abstract methods
 
     def prepare_launch(self, exec_path, args, file_to_open=None):
         """
         Prepares the given software for launch
 
         :param str exec_path: Path to DCC executable to launch
         :param str args: Command line arguments as strings
-        :param str file_to_open: (optional) Full path name of a file to open on launch
+        :param str file_to_open: (optional) Full path name of a file to open on
+            launch
         :returns: :class:`LaunchInformation` instance
         """
         raise NotImplementedError
 
+    def _format(self, template, tokens):
+        """
+        Limited implementation of Python 2.6-like str.format.
+
+        :param str template: String using {<name>} tokens for substitution.
+        :param dict tokens: Dictionary of <name> to substitute for <value>.
+
+        :returns: The substituted string, when "<name>" will yield "<value>".
+        """
+        for key, value in tokens.iteritems():
+            template = template.replace("{%s}" % key, value)
+        return template
+
+    def _glob_and_match(self, match_template, template_key_expressions):
+        """
+        This is a helper method that can be invoked in an implementation of :meth:`scan_software`.
+
+        The ``match_template`` argument provides a template to use both for globbing files and then pattern
+        matching them using regular expressions provided by the ``tokens_expressions`` dictionary.
+
+        The method will first substitute every token surrounded by `{}` from the template with a ``*``
+        for globbing files. It will then replace the tokens in the template with the regular expressions
+        that were provided.
+
+        In the following example::
+
+            self._glob_and_match(
+                "//network/softwares/Nuke{full_version}/Nuke{major_minor_version}.exe",
+                {
+                    "full_version": r"([\d.v]+)",
+                    "major_minor_version": r"([\d.]+)"
+                }
+            )
+
+        this would first look for every files matching C:/Program Files/Nuke*/Nuke*.exe and then
+        run the regular expression C:/Program Files/Nuke(?P<full_version>[\d.v]+)/(?P<major_minor_version>[\d.]+)
```

Comment:
Technically we're not running this expression. The path is different, and we're not assuming the `?P<>` part. It would be:

```
//network/softwares/Nuke([\d.v]+)/Nuke([\d.v]+).exe
```